jepsen/src/jepsen/checker.clj

```diff
@@ -359,57 +359,88 @@
+             ; Map: process ID -> list of [lower upper] pairs for the pending read operation invoked by the process.
+             ; Last pair in the list is most recent one and reflects the current counter possible range.
+             ; List is cleared once read operation is completed.
+             pending-reads      {}
+             ; Completed reads: list of [lower val upper] tuples - one tuple for each successful read.
+             ; Once read operation returns value val - we select first pair [lower upper] from pending-reads for which
+             ; lower <= val <= upper, or just the first pair if no such pair is found (that means we've found an
+             ; inconsistency, because val is out of any of possible ranges).
+             ; TODO: probably we can use single union range instead of list of ranges and that will makes reads and
+             ; errors reporting better.
+             reads              []]
+        (if (nil? history)
+          ; We're done here
+          (let [errors (remove (partial apply <=) reads)]
+            {:valid?             (empty? errors)
+             :reads              reads
+             :errors             errors})
+          ; But wait, there's more
+          (let [op      (first history)
+                process-id (:process op)
+                history (next history)]
+            (case [(:type op) (:f op)]
+              [:invoke :read]
+              (recur history lower upper
+                     (assoc pending-reads process-id [[lower upper]])
+                     reads)
+
+              [:ok :read]
+              (let [read-ranges (get pending-reads process-id)
+                    v (:value op)
+                    [l' u'] (first read-ranges)
+                    read (or (some (fn [[l u]] (when (<= l v u) [l v u])) read-ranges)
+                             [l' v u'])]
                 (recur history lower upper
-                       (assoc pending-reads (:process op) [lower (:value op)])
-                       reads)
-
-                [:ok :read]
-                (let [r (get pending-reads (:process op))]
-                  (recur history lower upper
-                         (dissoc pending-reads (:process op))
-                         (conj reads (conj r upper))))
-
-                [:invoke :add]
-                (recur history lower (+ upper (:value op)) pending-reads reads)
-
-                [:ok :add]
-                (recur history (+ lower (:value op)) upper pending-reads reads)
-
-                (recur history lower upper pending-reads reads))))))))
+                       (dissoc pending-reads process-id)
+                       (conj reads read)))
+
+              [:invoke :add]
+              (let [value (:value op)
+                    [l' u'] (if (> value 0) [lower (+ upper value)] [(+ lower value) upper])]
+                (recur history l' u'
+                       (reduce-kv #(assoc %1 %2 (conj %3 [l' u'])) {} pending-reads)
```

Comment:
I don't understand this.

I was expected something like 
(reduce-kv #(assoc %1 %2 [ (min (first %3) l') (max (second %3) u')])) {} pending-reads)

So, we are updating for each k, v (= [l u]) in pending-writes as  k -> v = [min (l, ,') max (u u')].

Reply:
It simply adds `[l' u']` range to list of ranges for each process in `pending-reads`:
```
jepsen.cli=> (let [
        #_=> upper 7 value -1 lower 3
        #_=> pending-reads {:1 [[1 2]] :2 [[2 4]]}
        #_=> [l' u'] (if (> value 0) [lower (+ upper value)] [(+ lower value) upper])]
        #_=> (reduce-kv #(assoc %1 %2 (conj %3 [l' u'])) {} pending-reads))
{:1 [[1 2] [2 7]], :2 [[2 4] [2 7]]}
```

I think this implementation which I simply ported from Cassandra's test suite could be improved, but for this iteration, it should be good enough.